master/internal/stream/supervisor.go

```
@@ -0,0 +1,150 @@
+}
+
+// monitorPermissionChanges listens for permission changes, updates the PublisherSet
+// to signal to boot streamers, returns an error in the event of a failure to listen.
+func monitorPermissionChanges(ctx context.Context, ps *PublisherSet) error {
+	permListener, err := AuthZProvider.Get().GetPermissionChangeListener()
+	if err != nil {
+		log.Errorf("unable to get permission change listener: %s", err)
+		// XXX: investigate better recovery mechanism
+		return err
+	}
+	if permListener == nil {
+		// no listener means we don't have permissions configured at all
+		return nil
+	}
+	defer func() {
+		err := permListener.Close()
+		if err != nil {
+			log.Debugf("error occurred while closing permission listener: %s", err)
+		}
+	}()
+
+	for {
+		select {
+		// did permissions change?
+		case <-permListener.Notify:
+			log.Debugf("permission change detected, booting streamers")
+			func() {
+				ps.permissionLock.Lock()
+				defer ps.permissionLock.Unlock()
+				close(ps.permissionChangeChan)
+				ps.permissionChangeChan = make(chan struct{})
+			}()
+		// is the listener still alive?
+		case <-time.After(30 * time.Second):
+			pingErrChan := make(chan error)
+			go func() {
+				err = permListener.Ping()
+				pingErrChan <- errors.Wrap(err, "no active connection")
+			}()
+			if err := <-pingErrChan; err != nil {
+				log.Errorf("permission listener failed %s", err)
+				return err
+			}
+		// are we canceled?
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (ssup *Supervisor) runOne(ctx context.Context) error {
+	group := errgroupx.WithContext(ctx)
+	subctx := group.Context()
+	func() {
+		ssup.cond.L.Lock()
+		defer ssup.cond.L.Unlock()
+		ssup.cond.Broadcast()
+
+		ssup.subctx = subctx
+		ssup.ps = NewPublisherSet()
+
+		// start monitoring permissions
+		group.Go(
+			func(c context.Context) error {
+				return monitorPermissionChanges(c, ssup.ps)
+			},
+		)
+		// start up all publishers
+		// XXX: should these be passing errors backup?
```

yes, absolutely

---

Currently `ps.Start()` eats it's own errors and continuously retries; I can add in retry logic similar to what's in supervisor (errgroupx + backoff). 

@rb-determined-ai any objections to swapping out the waitgroup for an errgroup in `ps.Start()`?

---

So goal here is to, use errgroupx & return the errors back up to the supervisor & have it decide to retry instead of having multiple levels of retry logic

---

exactly